### ISSUE TITLE
Support cancellation of movement

### ISSUE THREAD

**Issue by do0g** [bbc/lrud#25]
It is a common use case that some business logic needs to be performed after a move is initiated by the user but prior to the completion of the movement. The business logic may impact whether the completion of the movement occurs at all. Examples of this are failure to retrieve data, updating of internal state, validation that the initiated action is acceptable.

In these scenarios, some mechanism to allow the application to halt movement after it has been initiated and allow it to be completed after a successful business action has occurred would be ideal.

Possible solutions to this problem are:

- Being able to cancel a move operation, via the return of a boolean value from a move callback, or providing an API object method such as `.cancelEvent()`
- The move callback invoking the application handler with a callback function which could be invoked after the business logic is complete
- The move callback supporting promise-returning handlers, allowing continuation after the business logic is complete

**Comment by jordanholt** [comment]
Interesting use case... I think I would prefer the first described implementation where the move operation provides a callback, and if that callback returns true, the move completes.

**Comment by thomascgray** [comment]
Thanks for raising the issue @do0g , and for the chat earlier!

To continue from what Jordan said, some initial thoughts...

During the actual process of handling a movement, _before_ the `assignFocus()` is called (thus "completing the movement") we provide an opportunity for a callback to fire.

If said callback returns true, instead of assigning the focus and firing the `focus` and `move` events, we fire a `cancelled` event on the instance itself e.g 

```js
navigation.on('cancelled', (leaveNode, enterNode) => {
  // leaving `leaveNode` and entering `enterNode` was cancelled for reasons...
});
```

(We can also, at that point, call an `onCancelled` hook if one was supplied on the node to be focused.)

```js
navigation.registerNode('myBox', {
  parent: 'myContainer',
  onCancelled: (leave, enter) => {
    // movement from `leave` to `enter` was cancelled, where `enter` is _this_ node itself
  }
}
```

The actual cancellation callback _itself_ should _also_ live against the node that was to be "entered", I'm thinking? Some kind of `shouldCancel()` callback.

```js
navigation.registerNode('myBox', {
  parent: 'myContainer',
  onCancelled: (leave, enter) => {
    // movement from `leave` to `enter` was cancelled, where `enter` is _this_ node itself
  },
  shouldCancel: (leave, enter) : boolean => {
    // movement from `leave` to `enter` is happening, here is an opportunity
    // to call some business logic to `return false` and stop that movement
  }
}
```

Otherwise, perhaps a `shouldCancel()` on the instance itself? And it is called for every movement that happens. Before both in fact.

**Comment by jordanholt** [comment]
Another point worth thinking about is how this interacts (if at all) with the existing "overrides" mechanism.

**Comment by do0g** [comment]
This looks pretty good to me. Either of the node or instance solutions would probably work for us, but both would be very flexible.

**Comment by thomascgray** [comment]
Closing this ticket, as it was fixed by #30
